kaldb/src/main/java/com/slack/kaldb/server/OpenSearchBulkIngestAPI.java

```diff
@@ -0,0 +1,237 @@
+package com.slack.kaldb.server;
+
+import static com.linecorp.armeria.common.HttpStatus.INTERNAL_SERVER_ERROR;
+import static com.linecorp.armeria.common.HttpStatus.TOO_MANY_REQUESTS;
+import static com.slack.kaldb.metadata.dataset.DatasetMetadata.MATCH_ALL_SERVICE;
+import static com.slack.kaldb.metadata.dataset.DatasetMetadata.MATCH_STAR_SERVICE;
+import static com.slack.kaldb.preprocessor.PreprocessorService.CONFIG_RELOAD_TIMER;
+import static com.slack.kaldb.preprocessor.PreprocessorService.INITIALIZE_RATE_LIMIT_WARM;
+import static com.slack.kaldb.preprocessor.PreprocessorService.filterValidDatasetMetadata;
+import static com.slack.kaldb.preprocessor.PreprocessorService.sortDatasetsOnThroughput;
+
+import com.google.common.util.concurrent.AbstractService;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.server.annotation.Blocking;
+import com.linecorp.armeria.server.annotation.Post;
+import com.slack.kaldb.elasticsearchApi.BulkIngestResponse;
+import com.slack.kaldb.elasticsearchApi.OpenSearchRequest;
+import com.slack.kaldb.metadata.core.KaldbMetadataStoreChangeListener;
+import com.slack.kaldb.metadata.dataset.DatasetMetadata;
+import com.slack.kaldb.metadata.dataset.DatasetMetadataStore;
+import com.slack.kaldb.preprocessor.PreprocessorRateLimiter;
+import com.slack.kaldb.preprocessor.PreprocessorService;
+import com.slack.kaldb.proto.config.KaldbConfigs;
+import com.slack.service.murron.trace.Trace;
+import io.micrometer.core.instrument.Timer;
+import io.micrometer.prometheus.PrometheusMeterRegistry;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ThreadLocalRandom;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.errors.AuthorizationException;
+import org.apache.kafka.common.errors.OutOfOrderSequenceException;
+import org.apache.kafka.common.errors.ProducerFencedException;
+import org.apache.kafka.streams.kstream.Predicate;
+import org.opensearch.action.index.IndexRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OpenSearchBulkIngestAPI extends AbstractService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(OpenSearchBulkIngestAPI.class);
+  private final PrometheusMeterRegistry meterRegistry;
+
+  private final KaldbConfigs.PreprocessorConfig preprocessorConfig;
+  private final DatasetMetadataStore datasetMetadataStore;
+
+  private final KaldbMetadataStoreChangeListener<DatasetMetadata> datasetListener =
+      (datasetMetadata) -> load();
+
+  private final PreprocessorRateLimiter rateLimiter;
+  private Predicate<String, List<Trace.Span>> rateLimiterPredicate;
+  private List<DatasetMetadata> throughputSortedDatasets;
+
+  private final Timer configReloadTimer;
+
+  @Override
+  protected void doStart() {
```

Comment:
Does this need to be a guava service? It appears all we're doing is adding a listener to the metadata store?

Reply:
Currently the service on init does call `load` and also adds a listener. The listener is also removed on shutdown through the guava service API

This is very similar to what we do in the PreprocessorService. I thought guava service was helpful because it cleans up resources on shutdown etc